src/services/bunker.py

```diff
@@ -1,15 +1,344 @@
+        )
+        after_report_total_pooled_ether, _ = self.w3.lido_contracts.lido.functions.handleOracleReport(**args).call(
+            block_identifier=cdtp.ref_blockstamp.block_hash
+        )
+        logger.info({
+            "msg": "Simulate 'handleOracleReport' contract function call",
+            "call_args": args,
+            "before_call": before_report_total_pooled_ether,
+            "after_call": after_report_total_pooled_ether,
+        })
+
+        frame_cl_rebase = self.w3.from_wei(after_report_total_pooled_ether - before_report_total_pooled_ether, 'gwei')
+        logger.info({"msg": f"Simulated CL rebase for frame: {frame_cl_rebase} Gwei"})
+
+        return frame_cl_rebase
+
+    def _is_high_midterm_slashing_penalty(self, cdtp: CommonDataToProcess, cl_rebase: Gwei) -> bool:
+        logger.info({"msg": "Detecting high midterm slashing penalty"})
+        all_slashed_validators = self._not_withdrawn_slashed_validators(cdtp.ref_all_validators, cdtp.ref_epoch)
+        lido_slashed_validators: list[Validator] = [
+            v.validator for v in self.w3.lido_validators.filter_lido_validators(cdtp.lido_keys, all_slashed_validators)
+        ]
+        logger.info({"msg": f"Slashed: {len(all_slashed_validators)=} | {len(lido_slashed_validators)=}"})
+
+        # If no one lido in current slashed validators - no need to bunker
+        if not lido_slashed_validators:
+            return False
+
+        # Fill per epoch buckets by possible slashed epochs
+        per_epoch_buckets = defaultdict(list[Validator])
+        for validator in all_slashed_validators:
+            possible_epochs_with_slashings = self._detect_slashing_epoch_range(validator, cdtp.ref_epoch)
+            for epoch in possible_epochs_with_slashings:
+                per_epoch_buckets[epoch].append(validator)
+
+        min_bucket_epoch = min(per_epoch_buckets.keys())
+
+        # Iterate through per_epoch_buckets and calculate lido midterm penalties for each bucket
+        per_epoch_lido_midterm_penalties = defaultdict(int)
+        # We should calculate total_balance for each bucket, but we do it once for all per_epoch_buckets
+        total_balance = self._calculate_total_effective_balance(cdtp.ref_all_validators)
+        for epoch, slashed_validators in per_epoch_buckets.items():
+            slashed_keys = [k.validator.pubkey for k in slashed_validators]
+            lido_validators_slashed_in_epoch = [
+                v for v in lido_slashed_validators if v.validator.pubkey in slashed_keys
+            ]
+            if not lido_validators_slashed_in_epoch:
+                continue
+            # We should calculate penalties according to bounded slashings in past EPOCHS_PER_SLASHINGS_VECTOR
+            min_bounded_epoch = max(min_bucket_epoch, epoch - EPOCHS_PER_SLASHINGS_VECTOR)
+            bounded_slashings_balance_sum = sum([
+                sum(int(v.validator.effective_balance) for v in e_vals)
```

Comment:
you can use reduce here

Reply:
I'm not sure it's necessary. It's just sum of internal values by condition. it seems like we'll make readability worse this way

Reply:
reforged